### PR TITLE
rustdoc: use a trie for name-based search

### PR DESCRIPTION
**Pull request by notriddle**
Potentially https://github.com/rust-lang/rust/issues/131156 — need to try reproducing the problem with `windows`

Preview and profiler results
----------------------------

Here's some quick profiling in Firefox done on the rust compiler docs:

- Before: https://share.firefox.dev/3UPm3M8
- After: https://share.firefox.dev/40LXvYb

Here's the results for the node.js profiler:

- https://notriddle.com/rustdoc-html-demo-15/trie-perf/index.html

Here's a copy that you can use to try it out. Compare it with [the nightly]. Try typing `typecheckercontext` one character at a time, slowly.

- https://notriddle.com/rustdoc-html-demo-15/compiler-doc-trie/index.html

[the nightly]: https://doc.rust-lang.org/nightly/nightly-rustc/

The fuzzy match algo is based on [Fast String Correction with Levenshtein-Automata] and the corresponding implementation code in [moman] and [Lucene]; the bit-packing representation comes from Lucene, but the actual matcher is more based on `fsc.py`. As suggested in the paper, a trie is used to represent the FSA dictionary.

The same trie is used for prefix matching. Substring matching is done with a side table of three-character[^1] windows that point into the trie.

[Fast String Correction with Levenshtein-Automata]: https://github.com/tpn/pdfs/blob/master/Fast%20String%20Correction%20with%20Levenshtein-Automata%20(2002)%20(10.1.1.16.652).pdf
[Lucene]: https://fossies.org/linux/lucene/lucene/core/src/java/org/apache/lucene/util/automaton/Lev1TParametricDescription.java
[moman]: https://gitlab.com/notriddle/moman-rustdoc

User-visible changes
--------------------

I don't expect anybody to notice anything, but it does cause two changes:

- Substring matches, in the middle of a name, only apply if there's three or more characters in the search query.
- Levenshtein distance limit now maxes out at two. In the old version, the limit was w/3, so you could get looser matches for queries with 9 or more characters[^1] in them.
- It uses more RAM.
- It's faster (assuming you don't swap thrash).

[^1]: technically utf-16 code units

### CODE REVIEW

File: src/librustdoc/html/static/js/search.js [hunk 1]
```diff
@@ -1106,6 +1106,137 @@ class RoaringBitmapBits {
...
+        this.insertSubstring(name, 0, id, tailTable);
+    }
+    insertSubstring(name, substart, id, tailTable) {
+        const l = name.length;
+        if (substart === l) {
+            this.matches.push(id);
+        } else {
+            const sb = name.charCodeAt(substart);
+            let child;
+            if (this.children[sb] !== undefined) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1130]: Wouldn't it be better to check `this.children.length < sb`?

File: src/librustdoc/html/static/js/search.js [hunk 2]
```diff
@@ -1106,6 +1106,137 @@ class RoaringBitmapBits {
...
+                new Lev1TParametricDescription(name.length);
+            this.searchLev(name, 0, levParams, results);
+            const tail = name.substring(0, 3);
+            if (tailTable.has(tail)) {
+                for (const entry of tailTable.get(tail)) {
+                    entry.searchSubstringPrefix(name, 3, results);
+                }
+            } else {
+                console.log(tailTable);
+                console.log(tail);
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1165]: Forgot to remove logs?

File: src/librustdoc/html/static/js/search.js [hunk 3]
```diff
@@ -1106,6 +1106,137 @@ class RoaringBitmapBits {
...
+                }
+                if (unprocessedChildren.length === 0) {
+                    const tmp = unprocessedChildren;
+                    unprocessedChildren = nextSet;
+                    nextSet = tmp;
+                }
+            }
+        } else {
+            const sb = name.charCodeAt(substart);
+            if (this.children[sb] !== undefined) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1202]: Same here.

File: src/librustdoc/html/static/js/search.js (hunk 1 above)
Comment (notriddle) [reply src/librustdoc/html/static/js/search.js:1130]: This is a [sparse array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Indexed_collections#sparse_arrays).

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1130]: Please add this explanation and link on the field definition please. :)

Comment (notriddle) [reply src/librustdoc/html/static/js/search.js:1149]: Okay, it's added.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+

**Pull request merged**